File: code/java/ch16/src/main/java/com/github/zxh0/luago/compiler/parser/StatParser.java
package com.github.zxh0.luago.compiler.parser;

import com.github.zxh0.luago.compiler.ast.Block;
import com.github.zxh0.luago.compiler.ast.Exp;
import com.github.zxh0.luago.compiler.ast.Stat;
import com.github.zxh0.luago.compiler.ast.exps.*;
import com.github.zxh0.luago.compiler.ast.stats.*;
import com.github.zxh0.luago.compiler.lexer.Lexer;
import com.github.zxh0.luago.compiler.lexer.Token;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Map;

import static com.github.zxh0.luago.compiler.lexer.TokenKind.*;
import static com.github.zxh0.luago.compiler.parser.BlockParser.parseBlock;
import static com.github.zxh0.luago.compiler.parser.ExpParser.parseExp;
import static com.github.zxh0.luago.compiler.parser.ExpParser.parseExpList;
import static com.github.zxh0.luago.compiler.parser.ExpParser.parseFuncDefExp;
import static com.github.zxh0.luago.compiler.parser.PrefixExpParser.parsePrefixExp;

class StatParser {

    /*
    stat ::=  ‘;’
        | break
        | ‘::’ Name ‘::’
        | goto Name
        | do block end
        | while exp do block end
        | repeat block until exp
        | if exp then block {elseif exp then block} [else block] end
        | for Name ‘=’ exp ‘,’ exp [‘,’ exp] do block end
        | for namelist in explist do block end
        | function funcname funcbody
        | local function Name funcbody
        | local namelist [‘=’ explist]
        | varlist ‘=’ explist
        | functioncall
    */
    static Stat parseStat(Lexer lexer) {
        switch (lexer.LookAhead()) {
            case TOKEN_SEP_SEMI:    return parseEmptyStat(lexer);
            case TOKEN_KW_BREAK:    return parseBreakStat(lexer);
            case TOKEN_SEP_LABEL:   return parseLabelStat(lexer);
            case TOKEN_KW_GOTO:     return parseGotoStat(lexer);
            case TOKEN_KW_DO:       return parseDoStat(lexer);
            case TOKEN_KW_WHILE:    return parseWhileStat(lexer);
            case TOKEN_KW_REPEAT:   return parseRepeatStat(lexer);
            case TOKEN_KW_IF:       return parseIfStat(lexer);
            case TOKEN_KW_FOR:      return parseForStat(lexer);
            case TOKEN_KW_FUNCTION: return parseFuncDefStat(lexer);
            case TOKEN_KW_LOCAL:    return parseLocalAssignOrFuncDefStat(lexer);
            default:                return parseAssignOrFuncCallStat(lexer);
        }
    }

    // ;
    private static EmptyStat parseEmptyStat(Lexer lexer) {
        lexer.nextTokenOfKind(TOKEN_SEP_SEMI);
        return EmptyStat.INSTANCE;
    }

    // break
    private static BreakStat parseBreakStat(Lexer lexer) {
        lexer.nextTokenOfKind(TOKEN_KW_BREAK);
        return new BreakStat(lexer.line());
    }

    // ‘::’ Name ‘::’
    private static LabelStat parseLabelStat(Lexer lexer) {
        lexer.nextTokenOfKind(TOKEN_SEP_LABEL);          // ::
        String name = lexer.nextIdentifier().getValue(); // name
        lexer.nextTokenOfKind(TOKEN_SEP_LABEL);          // ::
        return new LabelStat(name);
    }

    // goto Name
    private static GotoStat parseGotoStat(Lexer lexer) {
        lexer.nextTokenOfKind(TOKEN_KW_GOTO);            // goto
        String name = lexer.nextIdentifier().getValue(); // name
        return new GotoStat(name);
    }

    // do block end
    private static DoStat parseDoStat(Lexer lexer) {
        lexer.nextTokenOfKind(TOKEN_KW_DO);  // do
        Block block = parseBlock(lexer);     // block
        lexer.nextTokenOfKind(TOKEN_KW_END); // end
        return new DoStat(block);
    }

    // while exp do block end
    private static WhileStat parseWhileStat(Lexer lexer) {
        lexer.nextTokenOfKind(TOKEN_KW_WHILE); // while
        Exp exp = parseExp(lexer);             // exp
        lexer.nextTokenOfKind(TOKEN_KW_DO);    // do
        Block block = parseBlock(lexer);       // block
        lexer.nextTokenOfKind(TOKEN_KW_END);   // end
        return new WhileStat(exp, block);
    }

    // repeat block until exp
    private static RepeatStat parseRepeatStat(Lexer lexer) {
        lexer.nextTokenOfKind(TOKEN_KW_REPEAT); // repeat
        Block block = parseBlock(lexer);        // block
        lexer.nextTokenOfKind(TOKEN_KW_UNTIL);  // until
        Exp exp = parseExp(lexer);              // exp
        return new RepeatStat(block, exp);
    }

    // if exp then block {elseif exp then block} [else block] end
    private static IfStat parseIfStat(Lexer lexer) {
        List<Exp> exps = new ArrayList<>();
        List<Block> blocks = new ArrayList<>();

        lexer.nextTokenOfKind(TOKEN_KW_IF);       // if
        exps.add(parseExp(lexer));                // exp
        lexer.nextTokenOfKind(TOKEN_KW_THEN);     // then
        blocks.add(parseBlock(lexer));            // block

        while (lexer.LookAhead() == TOKEN_KW_ELSEIF) {
            lexer.nextToken();                    // elseif
            exps.add(parseExp(lexer));            // exp
            lexer.nextTokenOfKind(TOKEN_KW_THEN); // then
            blocks.add(parseBlock(lexer));        // block
        }

        // else block => elseif true then block
        if (lexer.LookAhead() == TOKEN_KW_ELSE) {
            lexer.nextToken();                    // else
            exps.add(new TrueExp(lexer.line()));  //
            blocks.add(parseBlock(lexer));        // block
        }

        lexer.nextTokenOfKind(TOKEN_KW_END);      // end
        return new IfStat(exps, blocks);
    }

    // for Name ‘=’ exp ‘,’ exp [‘,’ exp] do block end
    // for namelist in explist do block end
    private static Stat parseForStat(Lexer lexer) {
        int lineOfFor = lexer.nextTokenOfKind(TOKEN_KW_FOR).getLine();
        String name = lexer.nextIdentifier().getValue();
        if (lexer.LookAhead() == TOKEN_OP_ASSIGN) {
            return finishForNumStat(lexer, name, lineOfFor);
        } else {
            return finishForInStat(lexer, name);
        }
    }

    // for Name ‘=’ exp ‘,’ exp [‘,’ exp] do block end
    private static ForNumStat finishForNumStat(Lexer lexer, String name, int lineOfFor) {
        ForNumStat stat = new ForNumStat();
        stat.setLineOfFor(lineOfFor);           // for
        stat.setVarName(name);                  // name

        lexer.nextTokenOfKind(TOKEN_OP_ASSIGN); // =
        stat.setInitExp(parseExp(lexer));       // exp
        lexer.nextTokenOfKind(TOKEN_SEP_COMMA); // ,
        stat.setLimitExp(parseExp(lexer));      // exp

        if (lexer.LookAhead() == TOKEN_SEP_COMMA) {
            lexer.nextToken();                  // ,
            stat.setStepExp(parseExp(lexer));   // exp
        } else {
            stat.setStepExp(new IntegerExp(lexer.line(), 1));
        }

        lexer.nextTokenOfKind(TOKEN_KW_DO);     // do
        stat.setLineOfDo(lexer.line());         //
        stat.setBlock(parseBlock(lexer));       // block
        lexer.nextTokenOfKind(TOKEN_KW_END);    // end

        return stat;
    }

    // for namelist in explist do block end
    // namelist ::= Name {‘,’ Name}
    // explist ::= exp {‘,’ exp}
    private static ForInStat finishForInStat(Lexer lexer, String name0) {
        ForInStat stat = new ForInStat();
                                                        // for
        stat.setNameList(finishNameList(lexer, name0)); // namelist
        lexer.nextTokenOfKind(TOKEN_KW_IN);             // in
        stat.setExpList(parseExpList(lexer));           // explist
        lexer.nextTokenOfKind(TOKEN_KW_DO);             // do
        stat.setLineOfDo(lexer.line());                 //
        stat.setBlock(parseBlock(lexer));               // block
        lexer.nextTokenOfKind(TOKEN_KW_END);            // end

        return stat;
    }

    // namelist ::= Name {‘,’ Name}
    private static List<String> finishNameList(Lexer lexer, String name0) {
        List<String> names = new ArrayList<>();
        names.add(name0);
        while (lexer.LookAhead() == TOKEN_SEP_COMMA) {
            lexer.nextToken();                            // ,
            names.add(lexer.nextIdentifier().getValue()); // Name
        }
        return names;
    }

    // local function Name funcbody
    // local namelist [‘=’ explist]
    private static Stat parseLocalAssignOrFuncDefStat(Lexer lexer) {
        lexer.nextTokenOfKind(TOKEN_KW_LOCAL);
        if (lexer.LookAhead() == TOKEN_KW_FUNCTION) {
            return finishLocalFuncDefStat(lexer);
        } else {
            return finishLocalVarDeclStat(lexer);
        }
    }

    /*
    http://www.lua.org/manual/5.3/manual.html#3.4.11

    function f() end          =>  f = function() end
    function t.a.b.c.f() end  =>  t.a.b.c.f = function() end
    function t.a.b.c:f() end  =>  t.a.b.c.f = function(self) end
    local function f() end    =>  local f; f = function() end

    The statement `local function f () body end`
    translates to `local f; f = function () body end`
    not to `local f = function () body end`
    (This only makes a difference when the body of the function
     contains references to f.)
    */
    // local function Name funcbody
    private static LocalFuncDefStat finishLocalFuncDefStat(Lexer lexer) {
        lexer.nextTokenOfKind(TOKEN_KW_FUNCTION);        // local function
        String name = lexer.nextIdentifier().getValue(); // name
        FuncDefExp fdExp = parseFuncDefExp(lexer);       // funcbody
        return new LocalFuncDefStat(name, fdExp);
    }

    // local namelist [‘=’ explist]
    private static LocalVarDeclStat finishLocalVarDeclStat(Lexer lexer) {
        String name0 = lexer.nextIdentifier().getValue();     // local Name
        List<String> nameList = finishNameList(lexer, name0); // { , Name }
        List<Exp> expList = null;
        if (lexer.LookAhead() == TOKEN_OP_ASSIGN) {
            lexer.nextToken();                                // ==
            expList = parseExpList(lexer);                    // explist
        }
        int lastLine = lexer.line();
        return new LocalVarDeclStat(lastLine, nameList, expList);
    }

    // varlist ‘=’ explist
    // functioncall
    private static Stat parseAssignOrFuncCallStat(Lexer lexer) {
        Exp prefixExp = parsePrefixExp(lexer);
        if (prefixExp instanceof FuncCallExp) {
            return new FuncCallStat((FuncCallExp) prefixExp);
        } else {
            return parseAssignStat(lexer, prefixExp);
        }
    }

    // varlist ‘=’ explist |
    private static AssignStat parseAssignStat(Lexer lexer, Exp var0) {
        List<Exp> varList = finishVarList(lexer, var0); // varlist
        lexer.nextTokenOfKind(TOKEN_OP_ASSIGN);         // =
        List<Exp> expList = parseExpList(lexer);        // explist
        int lastLine = lexer.line();
        return new AssignStat(lastLine, varList, expList);
    }

    // varlist ::= var {‘,’ var}
    private static List<Exp> finishVarList(Lexer lexer, Exp var0) {
        List<Exp> vars = new ArrayList<>();
        vars.add(checkVar(lexer, var0));               // var
        while (lexer.LookAhead() == TOKEN_SEP_COMMA) { // {
            lexer.nextToken();                         // ,
            Exp exp = parsePrefixExp(lexer);           // var
            vars.add(checkVar(lexer, exp));            //
        }                                              // }
        return vars;
    }

    // var ::=  Name | prefixexp ‘[’ exp ‘]’ | prefixexp ‘.’ Name
    private static Exp checkVar(Lexer lexer, Exp exp) {
        if (exp instanceof NameExp || exp instanceof TableAccessExp) {
            return exp;
        }
        lexer.nextTokenOfKind(null); // trigger error
        throw new RuntimeException("unreachable!");
    }

    // function funcname funcbody
    // funcname ::= Name {‘.’ Name} [‘:’ Name]
    // funcbody ::= ‘(’ [parlist] ‘)’ block end
    // parlist ::= namelist [‘,’ ‘...’] | ‘...’
    // namelist ::= Name {‘,’ Name}
    private static AssignStat parseFuncDefStat(Lexer lexer) {
        lexer.nextTokenOfKind(TOKEN_KW_FUNCTION);     // function
        Map<Exp, Boolean> map = parseFuncName(lexer); // funcname
        Exp fnExp = map.keySet().iterator().next();
        boolean hasColon = map.values().iterator().next();
        FuncDefExp fdExp = parseFuncDefExp(lexer);    // funcbody
        if (hasColon) { // insert self
            if (fdExp.getParList() == null) {
                fdExp.setParList(new ArrayList<>());
            }
            fdExp.getParList().add(0, "self");
        }

        return new AssignStat(fdExp.getLastLine(),
                Collections.singletonList(fnExp),
                Collections.singletonList(fdExp));
    }

    // funcname ::= Name {‘.’ Name} [‘:’ Name]
    private static Map<Exp, Boolean> parseFuncName(Lexer lexer) {
        Token id = lexer.nextIdentifier();
        Exp exp = new NameExp(id.getLine(), id.getValue());
        boolean hasColon = false;

        while (lexer.LookAhead() == TOKEN_SEP_DOT) {
            lexer.nextToken();
            id = lexer.nextIdentifier();
            Exp idx = new StringExp(id);
            exp = new TableAccessExp(id.getLine(), exp, idx);
        }
        if (lexer.LookAhead() == TOKEN_SEP_COLON) {
            lexer.nextToken();
            id = lexer.nextIdentifier();
            Exp idx = new StringExp(id);
            exp = new TableAccessExp(id.getLine(), exp, idx);
            hasColon = true;
        }

        // workaround: return multiple values
        return Collections.singletonMap(exp, hasColon);
    }

}

File: code/python/ch15_16/src/stat_parser.py
from lua_token import TokenKind
import lua_stat
import lua_exp
from exp_parser import ExpParser
from parser import Parser


"""
stat ::=  ‘;’
    | break
    | ‘::’ Name ‘::’
    | goto Name
    | do block end
    | while exp do block end
    | repeat block until exp
    | if exp then block {elseif exp then block} [else block] end
    | for Name ‘=’ exp ‘,’ exp [‘,’ exp] do block end
    | for namelist in explist do block end
    | function funcname funcbody
    | local function Name funcbody
    | local namelist [‘=’ explist]
    | varlist ‘=’ explist
    | functioncall
"""


class StatParser:
    @staticmethod
    def parse_stat(lexer):
        kind = lexer.look_ahead()
        if kind == TokenKind.SEP_SEMI:
            return StatParser.parse_empty_stat(lexer)
        if kind == TokenKind.KW_BREAK:
            return StatParser.parse_break_stat(lexer)
        if kind == TokenKind.SEP_LABEL:
            return StatParser.parse_label_stat(lexer)
        if kind == TokenKind.KW_GOTO:
            return StatParser.parse_goto_stat(lexer)
        if kind == TokenKind.KW_DO:
            return StatParser.parse_do_stat(lexer)
        if kind == TokenKind.KW_WHILE:
            return StatParser.parse_while_stat(lexer)
        if kind == TokenKind.KW_REPEAT:
            return StatParser.parse_repeat_stat(lexer)
        if kind == TokenKind.KW_IF:
            return StatParser.parse_if_stat(lexer)
        if kind == TokenKind.KW_FOR:
            return StatParser.parse_for_stat(lexer)
        if kind == TokenKind.KW_FUNCTION:
            return StatParser.parse_func_def_stat(lexer)
        if kind == TokenKind.KW_LOCAL:
            return StatParser.parse_local_assign_or_func_def_stat(lexer)
        return StatParser.parse_assign_or_func_call_stat(lexer)

    # ;
    @staticmethod
    def parse_empty_stat(lexer):
        lexer.get_next_token_of_kind(TokenKind.SEP_SEMI)
        return lua_stat.EmptyStat(lexer.get_line())

    # break
    @staticmethod
    def parse_break_stat(lexer):
        lexer.get_next_token_of_kind(TokenKind.KW_BREAK)
        return lua_stat.BreakStat(lexer.get_line())

    # '::' Name '::'
    @staticmethod
    def parse_label_stat(lexer):
        lexer.get_next_token_of_kind(TokenKind.SEP_LABEL)
        _, name = lexer.get_next_identifier()
        lexer.get_next_token_of_kind(TokenKind.SEP_LABEL)
        return lua_stat.LabelStat(name)

    # goto Name
    @staticmethod
    def parse_goto_stat(lexer):
        lexer.get_next_token_of_kind(TokenKind.KW_GOTO)
        _, name = lexer.get_next_identifier()
        return lua_stat.GotoStat(name)

    # do block end
    @staticmethod
    def parse_do_stat(lexer):
        lexer.get_next_token_of_kind(TokenKind.KW_DO)
        block = Parser.parse_block(lexer)
        lexer.get_next_token_of_kind(TokenKind.KW_END)
        return lua_stat.DoStat(block)

    # while exp do block end
    @staticmethod
    def parse_while_stat(lexer):
        lexer.get_next_token_of_kind(TokenKind.KW_WHILE)
        exp = ExpParser.parse_exp(lexer)
        lexer.get_next_token_of_kind(TokenKind.KW_DO)
        block = Parser.parse_block(lexer)
        lexer.get_next_token_of_kind(TokenKind.KW_END)
        return lua_stat.WhileStat(exp, block)

    # repeat block until exp
    @staticmethod
    def parse_repeat_stat(lexer):
        lexer.get_next_token_of_kind(TokenKind.KW_REPEAT)
        block = Parser.parse_block(lexer)
        lexer.get_next_token_of_kind(TokenKind.KW_UNTIL)
        exp = ExpParser.parse_exp(lexer)
        return lua_stat.RepeatStat(block, exp)

    # if exp then block {elseif exp then block} [else block] end
    @staticmethod
    def parse_if_stat(lexer):
        exps = []
        blocks = []

        lexer.get_next_token_of_kind(TokenKind.KW_IF)
        exps.append(ExpParser.parse_exp(lexer))
        lexer.get_next_token_of_kind(TokenKind.KW_THEN)
        blocks.append(Parser.parse_block(lexer))

        while lexer.look_ahead() == TokenKind.KW_ELSEIF:
            lexer.get_next_token()
            exps.append(ExpParser.parse_exp(lexer))
            lexer.get_next_token_of_kind(TokenKind.KW_THEN)
            blocks.append(Parser.parse_block(lexer))

        if lexer.look_ahead() == TokenKind.KW_ELSE:
            lexer.get_next_token()
            exps.append(lua_exp.TrueExp(lexer.get_line()))
            blocks.append(Parser.parse_block(lexer))

        lexer.get_next_token_of_kind(TokenKind.KW_END)
        return lua_stat.IfStat(exps, blocks)

    # for Name '=' exp ',' exp [',' exp] do block end
    # for namelist in explist do block end
    @staticmethod
    def parse_for_stat(lexer):
        line_of_for, _ = lexer.get_next_token_of_kind(TokenKind.KW_FOR)
        _, name = lexer.get_next_identifier()
        if lexer.look_ahead() == TokenKind.OP_ASSIGN:
            return StatParser.finish_for_num_stat(lexer, line_of_for, name)
        else:
            return StatParser.finish_for_in_stat(lexer, name)

    # for Name '=' exp ',' exp [',' exp] do block end
    @staticmethod
    def finish_for_num_stat(lexer, line_of_for, var_name):
        lexer.get_next_token_of_kind(TokenKind.OP_ASSIGN)
        init_exp = ExpParser.parse_exp(lexer)
        lexer.get_next_token_of_kind(TokenKind.SEP_COMMA)
        limit_exp = ExpParser.parse_exp(lexer)

        if lexer.look_ahead() == TokenKind.SEP_COMMA:
            lexer.get_next_token()
            step_exp = ExpParser.parse_exp(lexer)
        else:
            step_exp = lua_exp.IntegerExp(lexer.get_line(), 1)

        line_of_do, _ = lexer.get_next_token_of_kind(TokenKind.KW_DO)
        block = Parser.parse_block(lexer)
        lexer.get_next_token_of_kind(TokenKind.KW_END)

        return lua_stat.ForNumStat(line_of_for, line_of_do, var_name, init_exp, limit_exp, step_exp, block)

    # for Name '=' exp ',' exp [',' exp] do block end
    @staticmethod
    def finish_for_in_stat(lexer, name):
        name_list = StatParser.finish_name_list(lexer, name)
        lexer.get_next_token_of_kind(TokenKind.KW_IN)
        exp_list = ExpParser.parse_exp_list(lexer)
        line_of_do, _ = lexer.get_next_token_of_kind(TokenKind.KW_DO)
        block = Parser.parse_block(lexer)
        lexer.get_next_token_of_kind(TokenKind.KW_END)
        return lua_stat.ForInStat(line_of_do, name_list, exp_list, block)

    # namelist ::= Name {',' Name}
    @staticmethod
    def finish_name_list(lexer, name0):
        names = [name0]
        while lexer.look_ahead() == TokenKind.SEP_COMMA:
            lexer.get_next_token()
            _, name = lexer.get_next_identifier()
            names.append(name)
        return names

    #
    @staticmethod
    def parse_func_def_stat(lexer):
        lexer.get_next_token_of_kind(TokenKind.KW_FUNCTION)
        fn_exp, has_colon = StatParser.parse_func_name(lexer)
        fd_exp = ExpParser.parse_func_def_exp(lexer)
        if has_colon:
            fd_exp.par_list.insert(0, 'self')

        return lua_stat.AssignStat(fd_exp.line, fn_exp, fd_exp)

    # local function Name funcbody
    # local namelist ['=' explist]
    @staticmethod
    def parse_local_assign_or_func_def_stat(lexer):
        lexer.get_next_token_of_kind(TokenKind.KW_LOCAL)
        if lexer.look_ahead() == TokenKind.KW_FUNCTION:
            return StatParser.finish_local_func_def_stat(lexer)
        else:
            return StatParser.finish_local_var_decl_stat(lexer)

    """
    http: // www.lua.org / manual / 5.3 / manual.html  # 3.4.11

    function f() end = > f = function() end
    function t.a.b.c.f() end = > t.a.b.c.f = function() end
    function t.a.b.c: f() end = > t.a.b.c.f = function(self) end
    local function f() end = > local f; f = function() end

    The statement 'local function f() body end' 
    translates to 'local f; f = function() body end`
    not to `local f = function() body end`
    (This only makes a difference when the body of the function
    contains references to f.)
    """

    # local function Name funcbody
    @staticmethod
    def finish_local_func_def_stat(lexer):
        lexer.get_next_token_of_kind(TokenKind.KW_FUNCTION)
        _, name = lexer.get_next_identifier()
        fd_exp = ExpParser.parse_func_def_exp(lexer)
        return lua_stat.LocalFuncDefStat(name, fd_exp)

    # local namelist ['=' explist]
    @staticmethod
    def finish_local_var_decl_stat(lexer):
        _, name0 = lexer.get_next_identifier()
        name_list = StatParser.finish_name_list(lexer, name0)
        exp_list = []
        if lexer.look_ahead() == TokenKind.OP_ASSIGN:
            lexer.get_next_token()
            exp_list = ExpParser.parse_exp_list(lexer)
        last_line = lexer.get_line()
        return lua_stat.LocalVarDeclStat(last_line, name_list, exp_list)

    # varlist '=' explist
    # functioncall
    @staticmethod
    def parse_assign_or_func_call_stat(lexer):
        prefix_exp = ExpParser.parse_prefix_exp(lexer)
        if isinstance(prefix_exp, lua_exp.FuncCallExp):
            return prefix_exp
        else:
            return StatParser.parse_assign_stat(lexer, prefix_exp)

    # varlist '=' explist
    @staticmethod
    def parse_assign_stat(lexer, var0):
        var_list = StatParser.finish_var_list(lexer, var0)
        lexer.get_next_token_of_kind(TokenKind.OP_ASSIGN)
        exp_list = ExpParser.parse_exp_list(lexer)
        last_line = lexer.get_line()
        return lua_stat.AssignStat(last_line, var_list, exp_list)

    # varlist ::= var {',' var}
    @staticmethod
    def finish_var_list(lexer, var0):
        var_list = [StatParser.check_var(lexer, var0)]
        while lexer.look_ahead() == TokenKind.SEP_COMMA:
            lexer.get_next_token()
            exp = ExpParser.parse_prefix_exp(lexer)
            var_list.append(StatParser.check_var(lexer, exp))
        return var_list

    # var ::= Name | prefixexp '[' exp ']' | prefixexp '.' Name
    @staticmethod
    def check_var(lexer, exp):
        if isinstance(exp, lua_exp.NameExp) or isinstance(exp, lua_exp.TableAccessExp):
            return exp
        lexer.get_next_token_of_kind(-1)
        raise Exception('unreachable!')

    # function funcname funcbody
    # funcname ::= Name {'.' Name} [':' Name]
    # funcbody ::= '(' [parlist] ')' block end
    # parlist ::= namelist [',' '...'] | '...'
    # namelist ::= Name {',' Name}
    @staticmethod
    def parse_func_def_stat(lexer):
        lexer.get_next_token_of_kind(TokenKind.KW_FUNCTION)
        fn_exp, has_colon = StatParser.parse_func_name(lexer)
        fd_exp = ExpParser.parse_func_def_exp(lexer)
        if has_colon:
            fd_exp.insert(0, 'self')
        return lua_stat.AssignStat(fd_exp.line, [fd_exp], [fn_exp])

    # funcname ::= Name {'.' Name} [':' Name]
    @staticmethod
    def parse_func_name(lexer):
        line, name = lexer.get_next_identifier()
        exp = lua_exp.NameExp(line, name)
        has_colon = False

        while lexer.look_ahead() == TokenKind.SEP_DOT:
            lexer.get_next_token()
            line, name = lexer.get_next_identifier()
            idx = lua_exp.StringExp(line, name)
            exp = lua_exp.TableAccessExp(line, exp, idx)

        if lexer.look_ahead() == TokenKind.SEP_COLON:
            lexer.get_next_token()
            line, name = lexer.get_next_identifier()
            idx = lua_exp.StringExp(line, name)
            exp = lua_exp.TableAccessExp(line, exp, idx)
            has_colon = True

        return exp, has_colon
